README.md
# Storybook stylers.cloud Addon

A Storybook addon to use [stylers.cloud](https://www.stylers.cloud/?utm_source=gh) in your storybook

## Getting started

### 1. Install

Install using npm/yarn

```sh
yarn add -D storybook-addon-stylerscloud
```

### 2. Register the addon in `main.js`

To configure storybook to include addon, insert `storybook-addon-stylerscloud` to `addons` field in `.storybook/main.js`

```js
// .storybook/main.js
module.exports = {
  addons: ["storybook-addon-stylerscloud"],
};
```

### 3. Configure with project id

Configure addon with your stylers.cloud project id in `.storybook/preview.js` (edit or create it)

```js
// .storybook/preview.js
export const parameters = {
  stylersCloudProjectId: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
  stylersCloudNonProdDomains: [-1204607085, 1505998205],
});
```

You can find project id (`stylersCloudProjectId`) in client settings or in snippet we sent you.

If you want to use only [css.support](https://www.stylers.cloud/css-support/?utm_source=gh) this way add `stylersCloudSupportMode: true` to config and remove `stylersCloudNonProdDomains` (it is not needed).

#### Domain hashes

To get domain hashes required in `stylersCloudNonProdDomains` you can use this code:

1. open domain where stylers.cloud should run (eg. dev.example.com)
2. run code below in devtools console
3. output of script is hash of domain you are currently on (insert it to `stylersCloudNonProdDomains`)

```js
var domain = location.hostname.split(".").slice(-3).join(".");
function hash(s) {
  var h = 0,
    l = s.length,
    i;
  if (l == 0) return h;
  for (i = 0; i < l; i++) {
    h = (h << 5) - h + s.charCodeAt(i);
    h = h & h;
  }
  return h;
}
hash(domain);
```

exising defaults are

```
localhost = -1204607085
127.0.0.1 = 1505998205
```

## Using stylers.cloud in storybook

You can open stylers.cloud window from top-right corner inside Canvas/Docs frame.

![localhost_6006__path=_story_example-button--primary](https://user-images.githubusercontent.com/1194439/173340392-89e682ea-846e-4636-bc21-2c54b15c8169.png)

## More about stylers.cloud

[stylers.cloud](https://www.stylers.cloud/?utm_source=gh) is start-up automating CSS.
If you are not our client yet, feel free to [contact us](https://www.stylers.cloud/?utm_source=gh).


CHANGELOG.md
# v0.1.3 (Tue Aug 16 2022)

#### ⚠️ Pushed to `main`

- update lock file (pom@stylers.cloud)
- * v0.1.3 ([@seyd](https://github.com/seyd))
- * fix SDK to prod env ([@seyd](https://github.com/seyd))

#### Authors: 2

- Johnny Seyd ([@seyd](https://github.com/seyd))
- Stanislav Parnicky (pom@stylers.cloud)

---

# v0.1.3 (Tue Aug 16 2022)

#### ⚠️ Pushed to `main`

- * v0.1.3 ([@seyd](https://github.com/seyd))
- * fix SDK to prod env ([@seyd](https://github.com/seyd))

#### Authors: 1

- Johnny Seyd ([@seyd](https://github.com/seyd))

---

# v0.1.2 (Thu Jun 16 2022)

#### 🐛 Bug Fix

- fix: storybook addons catalog readme (pom@stylers.cloud)

#### ⚠️ Pushed to `main`

- docs: update readme (pom@stylers.cloud)

#### Authors: 1

- Stanislav Parnicky (pom@stylers.cloud)

---

# v0.1.1 (Thu Jun 16 2022)

#### 🐛 Bug Fix

- fix: using addon parameters ([@pomali](https://github.com/pomali))

#### Authors: 1

- [@pomali](https://github.com/pomali)

---

# v0.1.0 (Thu Jun 16 2022)

#### 🚀 Enhancement

- feat: bump + update readme ([@pomali](https://github.com/pomali))

#### ⚠️ Pushed to `main`

- docs: update readme ([@pomali](https://github.com/pomali))
- build: add conventional-commits auto plugin ([@pomali](https://github.com/pomali))

#### Authors: 1

- [@pomali](https://github.com/pomali)

---

# v0.0.6 (Wed Jun 15 2022)

#### ⚠️ Pushed to `main`

- project setup
- fix: update release to publish to npm (pom@stylers.cloud)
- docs: fix readme code formatting ([@pomali](https://github.com/pomali))
- docs: update readme ([@pomali](https://github.com/pomali))
- feat: add support for project mode ([@pomali](https://github.com/pomali))
- docs: update README.md ([@pomali](https://github.com/pomali))
- fix: update readme ([@pomali](https://github.com/pomali))
- fix: change readme npm -> yarn ([@pomali](https://github.com/pomali))
- fix: update package.json tags ([@pomali](https://github.com/pomali))
- Update README.md ([@pomali](https://github.com/pomali))
- feat: better readme, package metadata ([@pomali](https://github.com/pomali))
- fix: package.json ([@pomali](https://github.com/pomali))
- feat: remove unused code ([@pomali](https://github.com/pomali))
- feat: insert snippet into canvas ([@pomali](https://github.com/pomali))
- fix: decorator types ([@italoteix](https://github.com/italoteix))
- remove posinstall script after the first run ([@winkerVSbecks](https://github.com/winkerVSbecks))
- copy edits ([@winkerVSbecks](https://github.com/winkerVSbecks))
- add a welcome script ([@winkerVSbecks](https://github.com/winkerVSbecks))
- perf: Modify the corresponding ID ([@SCWR](https://github.com/SCWR))
- upgrade to storybook 6.4 ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Update README.md ([@literalpie](https://github.com/literalpie))
- Update Storybook from 6.2.9 to 6.3.6 ([@chantastic](https://github.com/chantastic))
- Fix issue with global persistently showing in URL ([@ghengeveld](https://github.com/ghengeveld))
- Update README.md ([@ghengeveld](https://github.com/ghengeveld))
- Add a prepublish check for README, use zx for all the scripts and fix deps ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Add essentials and fix selector for docs tab ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Use interface instead of type and React.FC ([@winkerVSbecks](https://github.com/winkerVSbecks))
- update README ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Add script to eject TS mode ([@winkerVSbecks](https://github.com/winkerVSbecks))
- migrate code to typescript ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Update withGlobals.js ([@kjenkins19](https://github.com/kjenkins19))
- upgrade deps ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Add both esm as a build target ([@winkerVSbecks](https://github.com/winkerVSbecks))
- switch to yarn ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Add dummy.ts to make tsc happy ([@shilman](https://github.com/shilman))
- Add out-of-the-box typescript support ([@shilman](https://github.com/shilman))
- Add a prerelease check for metadata ([@winkerVSbecks](https://github.com/winkerVSbecks))
- use dist/preset as main ([@winkerVSbecks](https://github.com/winkerVSbecks))
- start versioning at 0.0.0 ([@hipstersmoothie](https://github.com/hipstersmoothie))
- add automated release setup ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Add publishConfig for scoped packages ([@shilman](https://github.com/shilman))
- add lockfile ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Remove chromatic ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Add header and page stories ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Update README.md ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Panel expandable item and clear data when switching stories ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Add documentation ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Add more contextual info to the UI patterns ([@winkerVSbecks](https://github.com/winkerVSbecks))
- set up chromatic ([@winkerVSbecks](https://github.com/winkerVSbecks))
- extract presentational components ([@winkerVSbecks](https://github.com/winkerVSbecks))
- use a more realistic example of supportedFrameworks ([@winkerVSbecks](https://github.com/winkerVSbecks))
- use {} in template.bind ([@winkerVSbecks](https://github.com/winkerVSbecks))
- First commit ([@winkerVSbecks](https://github.com/winkerVSbecks))
- Create LICENSE ([@winkerVSbecks](https://github.com/winkerVSbecks))

#### Authors: 11

- [@pomali](https://github.com/pomali)
- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Benjamin Kindle ([@literalpie](https://github.com/literalpie))
- Gert Hengeveld ([@ghengeveld](https://github.com/ghengeveld))
- Ítalo Teixeira ([@italoteix](https://github.com/italoteix))
- Kyle Jenkins ([@kjenkins19](https://github.com/kjenkins19))
- Michael Chan ([@chantastic](https://github.com/chantastic))

src/preset/preview.ts
/**
 * A decorator is a way to wrap a story in extra “rendering” functionality. Many addons define decorators
 * in order to augment stories:
 * - with extra rendering
 * - gather details about how a story is rendered
 *
 * When writing stories, decorators are typically used to wrap stories with extra markup or context mocking.
 *
 * https://storybook.js.org/docs/react/writing-stories/decorators#gatsby-focus-wrapper
 */
import {
  PARAM_KEY_PROJECT_ID,
  PARAM_KEY_DOMAINS,
  PARAM_KEY_SUPPORT_MODE,
} from "../constants";
import { withGlobals } from "../withGlobals";

export const decorators = [withGlobals];

export const globalTypes = {
  [PARAM_KEY_PROJECT_ID]: {
    // "stylersCloudProjectId"
    name: PARAM_KEY_PROJECT_ID,
    description: "stylers.cloud project id",
    defaultValue: "please-fill-this-out",
  },
  [PARAM_KEY_DOMAINS]: {
    name: PARAM_KEY_DOMAINS,
    description: "stylers.cloud open on these domains (hashes)",
    defaultValue: [-1204607085, 1505998205],
  },
  [PARAM_KEY_SUPPORT_MODE]: {
    name: PARAM_KEY_SUPPORT_MODE,
    description: "use standalone mode",
    defaultValue: false,
  },
};


src/withGlobals.ts
import {
  DecoratorFunction,
  useParameter,
  useEffect,
  useGlobals,
} from "@storybook/addons";
import {
  PARAM_KEY_DOMAINS,
  PARAM_KEY_PROJECT_ID,
  PARAM_KEY_SUPPORT_MODE,
} from "./constants";

export const withGlobals: DecoratorFunction = (StoryFn, context) => {
  const projectId = useParameter(PARAM_KEY_PROJECT_ID);
  const domains = useParameter(PARAM_KEY_DOMAINS);
  const supportMode = useParameter(PARAM_KEY_SUPPORT_MODE);
  const stylersCloud = {
    [PARAM_KEY_PROJECT_ID]: projectId,
    [PARAM_KEY_DOMAINS]: domains,
    [PARAM_KEY_SUPPORT_MODE]: supportMode,
  };

  const globals = useGlobals();
  // Is the addon being used in the docs panel
  const isInDocs = context.viewMode === "docs";

  useEffect(() => {
    displayToolState({
      stylersCloud,
      globals,
      isInDocs,
    });
  }, [projectId]);

  return StoryFn();
};

function displayToolState(state: any) {
  if (!state.stylersCloud[PARAM_KEY_SUPPORT_MODE]) {
    const windowAny = window as any;
    if (windowAny.stylersCloud) return;
    var sc = (windowAny.stylersCloud = {
      config: {
        projectId: state.stylersCloud[PARAM_KEY_PROJECT_ID],
        nonProdDomains: state.stylersCloud[PARAM_KEY_DOMAINS] || [
          -1204607085, 1505998205,
        ], // see the link in comment above
        realtimePreview: true,
        nonFlickering: true,
      },
    });

    // @ts-ignore
    // prettier-ignore
    {function a(a){j.head.appendChild(a)}function b(a){var b=0,c=a.length,d;if(0==c)return b;for(d=0;d<c;d++){b=(b<<5)-b+a.charCodeAt(d);b=b&b}return b}function c(){return"loading"===j.readyState}var d="stylers.cloud",e=d.replace(".","-"),f="https://",g="script",h=sc.config,j=document,k=window.localStorage;if(k){var l=JSON.parse(k.getItem(e))||{};if(l.overrideConfig){for(var m in l){if(l.hasOwnProperty(m))h[m]=l[m]}}}var n=!h.env||3449687===b(h.env)?"":h.env+".";function o(){return f+n+"api."+d+"/v1/projects/"+h.projectId}function p(b,d){if(!d&&c()){j.write(b)}else{for(var e=b.split(/[><]/)[1].split(" "),f=j.createElement(e[0]),g=1,h;g<e.length;g++){h=e[g].split("=");f.setAttribute(h[0],h[1]?h[1].replace(/(^"|"$)/g,""):"")}a(f)}}var q=(!(99349===void 0)||h.nonFlickering)&&!h.supportModeOnly,r=h.realtimePreview&&!h.supportModeOnly;function s(){r&&p("<link rel=\"stylesheet\" type=\"text/css\" href=\""+o()+"/css?"+Date.now()+"\" "+e+"=\"global-css\">");q&&p("<"+g+" "+e+"=\""+"DOMagic"+"\" src=\""+o()+"/domodifier\"><"+"/"+g+">",!h.nonFlickering);p("<"+g+" src=\""+f+n+"sdk."+d+"/sdk.js\"><"+"/"+g+">",true)}var t=location.hostname.split(".").slice(-3).join(".");if(-1!==h.nonProdDomains.indexOf(b(t))){c()&&console.groupCollapsed(d+" snippet");c()&&console.log("Don't worry, the following warnings don't matter. Everything will work properly.");s();c()&&console.groupEnd()}else{h.realtimePreview=false;h.openAfterInit=true;sc.init=s}}
  } else {
    if (!(window as any).stylersCloud) {
      (window as any).stylersCloud = {
        config: {
          projectId: state.stylersCloud[PARAM_KEY_PROJECT_ID],
          supportModeOnly: true,
          openAfterInit: document.readyState === "complete",
        },
      };

      var script = document.createElement("script");
      script.src = "https://sdk.stylers.cloud/sdk.js";
      document.head.appendChild(script);
    }
  }
}
